BUG: Check paths are unicode, bytes or path-like

effigies opened the pull request:
In Python 2.7, `numpy.compat.py3k.os_fspath` checks `(str, bytes)`, which excludes `unicode` strings. This commit checks against `basestring`, which is PY2/3 compatible.

Fixes #12749

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

effigies committed:
BUG: Check paths are basestrings or path-like

In Python 2.7, numpy.compat.py3k.os_fspath checks (str, bytes), which
excludes unicode strings. This commit checks against basestring, which
is PY2/3 compatible.

Fixes #12749

eric-wieser reviewed (changes requested):
Pretty sure this now fails on 3, where basestring does not include bytes.

Just replace str with unicode here

effigies committed:
TEST: Verify os_fspath accepts all strings

effigies commented:
Should I also rebase to `maintenance/1.16.x` and change the merge target?

effigies committed:
FIX: Compare (unicode, bytes) over basestring

effigies commented:
I've also verified locally that the test I added fails prior to the patch and passes afterwards on Python 2.7, so it should serve as a regression test for that specific function. Is that sufficient, or should I make sure that some function that depends on this, like `numpy.save`, does not throw an error?

Finally, there's also this line further down the function:

https://github.com/numpy/numpy/blob/8d690f018c92f46d3b4d9d1132d30c326a60cdc1/numpy/compat/py3k.py#L238

I'm not sure if that can be hit in Python 2 with a `unicode` `path_repr`... It looks like it should always be a `str`, but let me know if you want me to update that line as well.

effigies commented:
@eric-wieser I hope I'm not breaking protocol to bump on this. Just want to note that if you're glancing over the green checks and red X's, the failure is just that coverage is no longer run for Python 2. This should be ready for another review.

charris commented:
Sounds like we only need this for 1.16, as Python 2.7 is not supported for later versions.

eric-wieser commented:
> but let me know if you want me to update that line as well.

Yes, you should do that

effigies commented:
> Sounds like we only need this for 1.16, as Python 2.7 is not supported for later versions.

Agreed.

effigies committed:
FIX: Compare (unicode, bytes) at later stage

effigies commented:
Removed 8a91ee9009e206fe461c816230d20abea59603a4 and rebased, because at least the test environment does not install `pathlib`. Let me know if you think it should be added and the test restored.

charris commented:
> basestring does not exist in Py3

NumPy has a compatible version: `from numpy.compat import basestring`
